test/core/prob_macro.jl

```diff
@@ -0,0 +1,78 @@
+using Turing, Distributions, Test, Random
+
+dir = splitdir(splitdir(pathof(Turing))[1])[1]
+include(dir*"/test/test_utils/AllUtils.jl")
+
+Random.seed!(129)
+
+@turing_testset "prob_macro" begin
+    @testset "scalar" begin
+        @model demo(x) = begin
+            m ~ Normal()
+            x ~ Normal(m, 1)
+        end
+
+        mval = 3
+        xval = 2
+        iters = 1000
+
+        logprior = logpdf(Normal(), mval)
+        loglike = logpdf(Normal(mval, 1), xval)
+        logjoint = logprior + loglike
+
+        @test logprob"m = mval | model = demo" == logprior
+        @test logprob"m = mval | x = xval, model = demo" == logprior
+        @test logprob"x = xval | m = mval, model = demo" == loglike
+        @test logprob"x = xval, m = mval | model = demo" == logjoint
+
+        varinfo = Turing.VarInfo(demo(xval))
+        @test logprob"m = mval | model = demo, varinfo = varinfo" == logprior
+        @test logprob"m = mval | x = xval, model = demo, varinfo = varinfo" == logprior
+        @test logprob"x = xval | m = mval, model = demo, varinfo = varinfo" == loglike
+        varinfo = Turing.VarInfo(demo(missing))
+        @test logprob"x = xval, m = mval | model = demo, varinfo = varinfo" == logjoint
+
+        chain = sample(demo(xval), IS(), iters)
+        chain2 = Chains(chain.value, chain.logevidence, chain.name_map, NamedTuple())
+        lps = logpdf.(Normal.(vec(chain["m"].value), 1), xval)
+        @test logprob"x = xval | chain = chain" == lps
+        @test logprob"x = xval | chain = chain2, model = demo" == lps
+        varinfo = Turing.VarInfo(demo(xval))
+        @test logprob"x = xval | chain = chain, varinfo = varinfo" == lps
+        @test logprob"x = xval | chain = chain2, model = demo, varinfo = varinfo" == lps
+    end
+
+    @testset "vector" begin
+        n = 5
+        @model demo(x, n = n, ::Type{T} = Float64) where {T} = begin
+            m = Vector{T}(undef, n)
+            @. m ~ Normal()
+            @. x ~ Normal.(m, 1)
+        end
+        mval = rand(n)
+        xval = rand(n)
+        iters = 1000
+
+        logprior = sum(logpdf.(Normal(), mval))
+        like(m, x) = sum(logpdf.(Normal.(m, 1), x))
+        loglike = like(mval, xval)
+        logjoint = logprior + loglike
+
+        @test logprob"m = mval | model = demo" == logprior
+        @test logprob"x = xval | m = mval, model = demo" == loglike
+        @test logprob"x = xval, m = mval | model = demo" == logjoint
+
+        varinfo = Turing.VarInfo(demo(xval))
+        @test logprob"m = mval | model = demo, varinfo = varinfo" == logprior
+        @test logprob"x = xval | m = mval, model = demo, varinfo = varinfo" == loglike
```

Comment:
In my understanding this is equivalent to: `p(x = xval | m = mval, model = demo, varinfo = varinfo) = p(x = xval | m = mval, model = demo, x = xval) = p(x = xval | m = mval, model = demo)`. Is this correct? If so, I feel there should be a comment explaining this here.

Additional test:
* Test that `logprob"x = xval | model = demo, varinfo = varinfo" == loglike` if `m = mval` in varinfo

Reply:
* Test that `logprob"x = xval | model = demo, m = 0, varinfo = varinfo"` throws an error if `m = mval != 0` in varinfo

Reply:
Well in the above syntax, `varinfo` is only for pre-allocation and type stability. If this is done for many values of `xval` in a loop, having a single `varinfo` is necessary for performance. I will add this to the docs. So the value inside `varinfo` is overwritten.

Reply:
I see. In that case I find the syntax confusing. Would it be possible to do this differently. Currently it implies to me that we condition on VarInfo, which we don’t actually do if I understand it correctly.

Reply:
True, do you have an alternative in mind?

Reply:
Perhaps use `;` instead of `,`?

Reply:
I take that back :)

Reply:
Let's open an issue and deal with this in a separate PR. What do you think?

Reply:
https://github.com/TuringLang/Turing.jl/issues/1038